Refactor the accumulator module.

busykoala opened the pull request:
🚧 Do not merge yet 🚧 

close #79 

I tried to do little steps for the refactoring by splitting it into different commits. Maybe not all of them make sense on their own but will if they are in larger context (especially I tried to minimize the diff so that one can tell what actually changed and what was only moved).

Commit:
Extract link search on fields into own class.

I moved all code responsible for link search in fields into its own class to split different task into different units. Also I was able to reduce the complexity of the unittests because now I don't need the platform anymore (when instancing a LinkOnFieldSeeker instance).

Since the module does not only contain the accumulator from now on I tried to give it a better fitting name as well.

Commit:
Move AT/DX field search into own methods.

This commit copies the code from within the condition (AT/DX) into
methods. Other than that I only removed the public interface method
since it was only calling the other one (which is now serving as the
pub. interface method).

I tried to reduce the complexity and readability by splitting the code (within the condition choosing between AT and DX) into their own functions.

busykoala commented:
I hope the refactoring done helps understand the code better. It still did not touch how things are done actually but I tried to split the code by functionality. Having methods with descriptive names helps to understand what the steps actually do.

mbaechtold commented on ftw/linkchecker/link_accumulation.py:
```diff
@@ -47,7 +47,8 @@ def _get_all_links(self):
         link_objs = []
         for brain in brains:
-            link_objs.extend(self._find_links_on_brain_fields(brain))
+            link_seeker = LinkOnFieldSeeker()
```
Why do you need a new link seeker for each brain? Since you pass the brain to the link seeker, the link seeker could be reused?

busykoala replied:
Great catch! I moved it upfront the for-loop now (in c9d9920) so it will only be instantiated once per Accumulator instance.

mbaechtold commented on ftw/linkchecker/link_accumulation.py:
```diff
@@ -61,64 +62,75 @@ def _separate_links_to_internal_external(self):
 
-    def _find_links_on_brain_fields(self, brain):
+
+class LinkOnFieldSeeker(object):
```
The name of the class is hard to grasp for me.

A docstring would really be helpful in order to understand what a LinkOnFieldSeeker does: 
- Does it extract links from brains?
- And where does the word "field" from the name of the class come into play?

busykoala replied:
True point, thanks for mentioning. I tried to summarize the steps done in a docstring (in c9d9920). Hopefully, that helps to find the code parts better responsible for each step. I was thinking of also referencing and documenting each method but I'm not sure if it improves the code really. What do you think?

busykoala replied:
b928a08 unternimmt einen weiteren Versuch etwas besser zu erklären.

mbaechtold commented on ftw/linkchecker/link_accumulation.py:
```diff
@@ -61,64 +62,75 @@ def _separate_links_to_internal_external(self):
+
+        return link_objs
+
+    def _find_links_on_archetypes_brain_fields(self, obj, link_objs):
```
I know, it's just refactoring, but this method is quite long and has too many levels of indentation for my taste. Maybe you could consider to refactor this into two separate helper classes and a base class sharing some common code. The seeker could identify the type (AT/DX) and then ask the appropriate class to extract the links?

busykoala replied:
In 84e23f8 I tried to split the LinkOnFieldSeeker further up. It instantiates either an ArchetypeSeeker or a DexteritySeeker which inherit from the BaseSeeker class (containing most of the methods used).
In a further commit, I will try to document more correct what they do.

I will look at the method mentioned. Maybe I'm going to do that as well later on.

busykoala replied:
I tried to describe the task of every class in this module in b928a08. I hope the understanding was improved a little.

mbaechtold reviewed (commented):
Just a few comments.

Commit:
Instantiate LinkOnFieldSeeker only once.

By moving the instantiating of LinkOnFieldSeeker out of the for loop it will only be instantiated once per Accumulator instance.

Commit:
Add docstring for LinkOnFieldSeeker.

I added the docstring to help understand what the class does.

Commit:
Put code in different classes.

This commit seems like a lot of change. What I did is actually only moving around code and renaming methods. The structure now changed to a LinkOnFieldSeeker which either instantiates a Archetype or Dexterity Seeker. Both of the last two inherit from BaseSeeker which contains the shared methods.

Commit:
Improve/add docstrings for classes in link_accumulation.

I tried to describe what the task of every class is. Hopefully that helps to
create a better understanding to further improve this module.

mbaechtold reviewed (approved):
Looks good to me 👍

Commit:
Split up the public method of ArchetypeSeeker.

busykoala commented:
I split up the public method of ArchetypeSeeker (the one you suggested) in cb1ab9e.
At least for me, it makes more sense right now the way I changed it ;-)

Commit:
Add changelog.
